cookbooks/inner_core_convection/inner_core_convection.cc

```diff
@@ -0,0 +1,424 @@
+      // product of density and specific heat a constant.
+      Simple<dim>::evaluate(in, out);
+
+      // We want the right-hand side of the momentum equation to be (- Ra T gravity) and
+      // density * cp to be 1
+      for (unsigned int q=0; q < in.position.size(); ++q)
+        {
+          out.densities[q] = - out.thermal_expansion_coefficients[q] * (in.temperature[q]);
+          if (std::abs(out.densities[q]) > 0.0)
+            out.specific_heat[q] /= out.densities[q];
+        }
+    }
+
+    template <int dim>
+    InnerCore<dim>::InnerCore ()
+      :
+      resistance_to_phase_change (1)
+    {}
+
+
+    template <int dim>
+    void
+    InnerCore<dim>::update()
+    {
+      // we get time passed as seconds (always) but may want
+      // to reinterpret it in years
+      if (this->convert_output_to_years())
+        resistance_to_phase_change.set_time (this->get_time() / year_in_seconds);
+      else
+        resistance_to_phase_change.set_time (this->get_time());
+    }
+
+
+    template <int dim>
+    void
+    InnerCore<dim>::declare_parameters (ParameterHandler &prm)
+    {
+      Simple<dim>::declare_parameters (prm);
+
+      prm.enter_subsection("Material model");
+      {
+        prm.enter_subsection("Inner core");
+        {
+          prm.enter_subsection("Phase change resistance function");
+          {
+            Functions::ParsedFunction<dim>::declare_parameters (prm, 1);
+          }
+          prm.leave_subsection();
+        }
+        prm.leave_subsection();
+      }
+      prm.leave_subsection();
+    }
+
+
+    template <int dim>
+    void
+    InnerCore<dim>::parse_parameters (ParameterHandler &prm)
+    {
+      Simple<dim>::parse_parameters (prm);
+
+      prm.enter_subsection("Material model");
+      {
+        prm.enter_subsection("Inner core");
+        {
+          prm.enter_subsection("Phase change resistance function");
+          try
+            {
+              resistance_to_phase_change.parse_parameters (prm);
+            }
+          catch (...)
+            {
+              std::cerr << "ERROR: FunctionParser failed to parse\n"
+                        << "\t'Phase boundary model.Function'\n"
+                        << "with expression\n"
+                        << "\t'" << prm.get("Function expression") << "'"
+                        << "More information about the cause of the parse error \n"
+                        << "is shown below.\n";
+              throw;
+            }
+          prm.leave_subsection();
+        }
+        prm.leave_subsection();
+      }
+      prm.leave_subsection();
+    }
+
+  }
+}
+
+
+
+namespace aspect
+{
+  namespace HeatingModel
+  {
+    using namespace dealii;
+
+    /**
+     * A class that implements a constant radiogenic heating rate.
+     *
+     * @ingroup HeatingModels
+     */
+    template <int dim>
+    class ConstantCoreHeating : public Interface<dim>
```

Comment:
Once we have the approximation for BA/EBA in the main code, this won't be necessary any more.